### PR TITLE
per-client settings

### PR DESCRIPTION
**Pull request by janisozaur**
This allows per-client settings, such as "disable clearance checks". I must've made a mistake somewhere, as it doesn't work yet, but if possible, I would like it to get included in the upcoming release.

### CODE REVIEW

File: src/network/NetworkPlayer.h [hunk 1]
```diff
@@ -40,6 +41,7 @@ class NetworkPlayer
     uint32      last_action_time    = 0;
     rct_xyz16   last_action_coord   = { 0 };
     std::string keyhash;
+	std::map<int, uintptr_t> user_prefs;
```
Comment (IntelOrca) [inline src/network/NetworkPlayer.h]: tab -> space

File: src/network/NetworkPlayer.h [hunk 2]
```diff
@@ -17,6 +17,7 @@
 #pragma once
 
 #include <string>
+#include <map>
```
Comment (IntelOrca) [inline src/network/NetworkPlayer.h]: Order includes of same type alphabetically.

File: src/network/network.cpp [hunk 3]
```diff
@@ -2367,6 +2403,41 @@ const utf8 * network_get_server_provider_name() { return gNetwork.ServerProvider
...
+	if (player == nullptr) {
+		return;
+	}
+	for (const auto &it : player->user_prefs) {
+		switch (it.first) {
+			case CLIENT_PREFS::CLIENT_PREF_DISABLE_CLEARENCE:
+				// Even if a server does not allow using disable clearance and so this cannot be *set*,
+				// a client can still have this setting on, so we need to check anyway.
+				if (gServerCheatsDisableClearanceChecks) {
+					gCheatsDisableClearanceChecks = !!(it.second);
```
Comment (IntelOrca) [inline src/network/network.cpp]: `!!`, this like the sort cryptic syntax you see in javascript. Why not `it.second != 0`?

File: src/rct2/S6Exporter.cpp [hunk 4]
```diff
@@ -545,7 +545,7 @@ extern "C"
         SDL_WriteLE32(rw, _guestGenerationProbability);
         SDL_WriteLE32(rw, _suggestedGuestMaximum);
         SDL_WriteU8(rw, gCheatsSandboxMode);
-        SDL_WriteU8(rw, gCheatsDisableClearanceChecks);
+		SDL_WriteU8(rw, gServerCheatsDisableClearanceChecks);
```
Comment (IntelOrca) [inline src/rct2/S6Exporter.cpp]: Tabs -> spaces

File: src/rct2/S6Importer.cpp [hunk 5]
```diff
@@ -475,7 +475,7 @@ extern "C"
         _guestGenerationProbability = SDL_ReadLE32(rw);
         _suggestedGuestMaximum = SDL_ReadLE32(rw);
         gCheatsSandboxMode = SDL_ReadU8(rw) != 0;
-        gCheatsDisableClearanceChecks = SDL_ReadU8(rw) != 0;
+		gServerCheatsDisableClearanceChecks = SDL_ReadU8(rw) != 0;
```
Comment (IntelOrca) [inline src/rct2/S6Importer.cpp]: Tabs -> spaces

File: src/localisation/string_ids.h [hunk 6]
```diff
@@ -3572,6 +3572,8 @@ enum {
 	STR_COMMA16 = 5919,
 	STR_RENDER_WEATHER_EFFECTS = 5920,
 	STR_RENDER_WEATHER_EFFECTS_TIP = 5921,
+	STR_ALLOW_USING_DISABLE_CLEARANCE = 5922,
+	STR_ACTION_ALLOW_ADJUSTING_CLIENT_PREFS = 5923,
```
Comment (IntelOrca) [inline src/localisation/string_ids.h]: There is no string for this.

**Comment by IntelOrca** [comment]
I can't enable the permission for clearance on a new group because the admin group doesn't have the permission by default.

**Comment by IntelOrca** [comment]
When 'allow using disable clearance checks' is enabled in cheats, it allows it even though its not enabled via the top menu.

File: src/network/network.cpp [hunk 7]
```diff
@@ -2237,6 +2237,42 @@ void game_command_kick_player(int *eax, int *ebx, int *ecx, int *edx, int *esi,
 	*ebx = 0;
 }
 
+void game_command_adjust_client_prefs(int *eax, int *ebx, int *ecx, int *edx, int *esi, int *edi, int *ebp)
+{
+	uint8 playerid = (uint8)*eax;
```
Comment (duncanspumpkin) [inline src/network/network.cpp]: It is always much nicer to the next person if you make the first part of your game command assignments of all of the register pointers. I.e. I want to see uint8 flags = *ebx, userprefs = *edx, ...

**Comment by janisozaur** [comment]
I have updated it a little bit, one major thing that is still missing is ability for server to use it, when that seat is in use (host also playing the game) and somewhat smaller one: server needs to validate player's request, so he can only change his prefs.

**Comment by janisozaur** [comment]
Updated. This should actually work now. 🎉

**Comment by IntelOrca** [comment]
This needs rebasing, string ID conflict.

**Comment by janisozaur** [comment]
rebased

**Comment by IntelOrca** [comment]
"Allow using disable clearance checks" should be the permission, not "Allow clients to modify their settings"

**Comment by janisozaur** [comment]
There could be more per-client settings than just this one.

**Comment by IntelOrca** [comment]
@janisozaur a client can always change their client settings, but only ones they have permission for. Why would they otherwise not be able to? There should be a permission for each client setting they can use.

**Comment by janisozaur** [comment]
In that case there's no need to use a map for storing these settings but rather add a dedicated field in player class for each of them.

**Comment by IntelOrca** [comment]
That would be one solution, ideally it should be flags for sandbox mode, disable clearance heights etc. passed into the game command.

**Comment by janisozaur** [comment]
updated

**Comment by janisozaur** [comment]
This is updated now

**Comment by janisozaur** [comment]
I'll take another stab at it at some point.

**Pull request closed without merging**